extensions/amp-image-slider/1.0/component.js

```diff
@@ -0,0 +1,552 @@
+import objstr from 'obj-str';
+
+import {setStyle} from '#core/dom/style';
+import {clamp} from '#core/math';
+
+import * as Preact from '#preact';
+import {
+  useCallback,
+  useEffect,
+  useImperativeHandle,
+  useLayoutEffect,
+  useMemo,
+  useRef,
+  useState,
+} from '#preact';
+import {forwardRef} from '#preact/compat';
+
+import {useStyles} from './component.jss';
+
+import {listen} from '../../../src/event-helper';
+import {Gestures} from '../../../src/gesture';
+import {SwipeXRecognizer} from '../../../src/gesture-recognizers';
+
+const VALID_IMAGE_TAGNAMES = new Set(['AMP-IMG', 'IMG']);
+
+/**
+ * Displays given component with supplied props.
+ * @param {*} component Component to render
+ * @return {PreactDef.Renderable}
+ */
+function DisplayAs({as: Comp = 'div', ...rest}) {
+  return <Comp {...rest} />;
+}
```

Comment:
Hi @caroqliu,

I am trying to assign reference to the dynamic images through `<DisplayAs .../>` component:
```jsx
/**
 * Displays given component with supplied props.
 * @param {*} props
 * @param {{current: ?Element}} ref
 * @return {PreactDef.Renderable}
 */
function DisplayAsWithRef({as: Comp = 'div', ...rest}, ref) {
  return <Comp {...rest} ref={ref} />;
}

const DisplayAs = forwardRef(DisplayAsWithRef);

...

// Not working ❌
<DisplayAs as={firstImageAs} ref={leftImageRef} />

// This is working ✅
<img
  src="https://amp.dev/static/samples/img/canoe_900x600_blur.jpg"
  ref={leftImageRef}
  style={{width: 600, height: 300}}
/>
```
Reference inside `DisplayAs` component remains `undefined`.

I am referring [`<ContainWrapper .../>`](https://github.com/ampproject/amphtml/blob/09a67efa5446ebd33c03d7fb0f15789609ce0930/src/preact/component/contain.js
)
What am I missing?

Reply:
I'm not sure what you mean, can you describe what you are expecting and what is not occuring when you say it is not working? I see that if I `console.log(ref)` inside `DisplayAsWithRef`, `leftImageRef` is defined as `{current: null|Node}`. Are you accounting for that the first render will leave `ref` as `{current: null}`, but second render it will be defined?

Reply:
Actually, `leftImageRef` and `rightImageRef` both remains as `null` all the time while using `<DisplayAs .../>`.

I am passing reference to display as:
```jsx
<DisplayAs as={firstImageAs} ref={leftImageRef} />
```

These `null` reference creates problem @ https://github.com/rtCamp/amphtml/blob/bdcb443e1bf3c004b34063d3c1ecf318faee9f66/extensions/amp-image-slider/1.0/component.js#L244:
```jsx
/**
 * Update element positions based on percentage
 * Should be wrapped inside mutateElement
 * @param {number} percentFromLeft
 * @private
 */
const updatePositions = useCallback(
  (percentFromLeft) => {
    ...
    updateTranslateX(rightImageRef.current, -percentFromLeft);   // < -- rightImageRef.current remains NULL
    ...
  },
  [limitPercentage, updateTranslateX]
);
```


Sample video of using reference directly as well as with `<DisplayAs .../>` component:
https://user-images.githubusercontent.com/75766877/135639056-5a72dfaf-811c-4f6c-93c1-50eb0688e79a.mp4

I am not sure what you mean by first and second render:
> Are you accounting for that the first render will leave ref as null, but second render it will be defined?

Reply:
Thank you for sending the video. I am seeing an entirely different UI when running your branch locally... 
![Screen Shot 2021-10-01 at 1 21 37 PM](https://user-images.githubusercontent.com/10456171/135661714-11bd5ce0-d523-4d58-973c-97a44c1bdb4c.png)

We should look into why that is separately. Taking a look at the ref now.

Reply:
Regarding the `ref` being `null`, I see that the issue is with the Storybook example which ignores the ref. Updating the sample to the following makes the ref defined:
```js
firstImageAs={forwardRef((props, ref) => (
  <img ref={ref} src={first} alt="A green apple" {...props} />
))}
```

However, since this isn't particularly ergonomic for developers, what do you think about wrapping the image inside a div with the ref for a simpler devX (will not have to use the above)?
```js
function DisplayAsWithRef({as: Comp = 'div', ...rest}, ref) {
  // this hook can be removed, but demonstrates that the ref will be defined after render
  useEffect(() => {
    console.log(ref.current);
  }, [ref]);
  return (
    <div ref={ref}>
      <Comp {...rest} />
    </div>
  );
}
```

Regarding the first render phase - the first time the return jsx is rendered in the dom, the given `ref` will be `null` because it relies on the dom references - prior to the render occurring, this cannot be defined. So it is not until after the first render is complete that `ref.current` will be defined.

Reply:
> Thank you for sending the video. I am seeing an entirely different UI when running your branch locally... ![Screen Shot 2021-10-01 at 1 21 37 PM](https://user-images.githubusercontent.com/10456171/135661714-11bd5ce0-d523-4d58-973c-97a44c1bdb4c.png)
> 
> We should look into why that is separately. Taking a look at the ref now.

I am really sorry. I didn't pushed experimental updates to this branch.
I have just pushed experimental commit so you can experiment: https://github.com/ampproject/amphtml/pull/35783/commits/4d8699bd65a2a67cb7269185f7a25015a350959b

Please let me know if you found any issue/correction.

Reply:
> Regarding the `ref` being `null`, I see that the issue is with the Storybook example which ignores the ref. Updating the sample to the following makes the ref defined:
> 
> ```js
> firstImageAs={forwardRef((props, ref) => (
>   <img ref={ref} src={first} alt="A green apple" {...props} />
> ))}
> ```
> 
> However, since this isn't particularly ergonomic for developers, what do you think about wrapping the image inside a div with the ref for a simpler devX (will not have to use the above)?
> 
> ```js
> function DisplayAsWithRef({as: Comp = 'div', ...rest}, ref) {
>   // this hook can be removed, but demonstrates that the ref will be defined after render
>   useEffect(() => {
>     console.log(ref.current);
>   }, [ref]);
>   return (
>     <div ref={ref}>
>       <Comp {...rest} />
>     </div>
>   );
> }
> ```
> 
> Regarding the first render phase - the first time the return jsx is rendered in the dom, the given `ref` will be `null` because it relies on the dom references - prior to the render occurring, this cannot be defined. So it is not until after the first render is complete that `ref.current` will be defined.

I think this is the solution I was looking for💡
Let me experiment and will let you know result 🙂